feat(group_theory): Burnside lemma

urkud opened the pull request:
TO CONTRIBUTORS:

Make sure you have:

  * [X] reviewed and applied the coding style: [coding](https://github.com/leanprover/mathlib/blob/master/docs/contribute/style.md), [naming](https://github.com/leanprover/mathlib/blob/master/docs/contribute/naming.md)
  * [ ] for tactics:
     * [ ] added or adapted documentation in [tactics.md](https://github.com/leanprover/mathlib/blob/master/docs/tactics.md)
     * [ ] write an example of use of the new feature in [tactics.lean](https://github.com/leanprover/mathlib/blob/master/tests/tactics.lean)
  * [X] make sure definitions and lemmas are put in the right files: thinking about correct abstraction levels for various facts made me add a few new files.
  * [X] make sure definitions and lemmas are not redundant

If this PR is related to a discussion on Zulip, please include a link in the discussion.

For reviewers: [code review check list](https://github.com/leanprover/mathlib/blob/master/docs/contribute/code-review.md)

urkud committed:
chore(group_theory/group_action): split `orbit_equiv_quotient_stabilizer`

Reasons:

* most of the steps are computable
* some facts are useful on their own;
    e.g., `g • b = h • b` iff `g` and `h` are in the same left coset

urkud committed:
feat(group_theory): Prove Burnside lemma

Probably, some lemmas should find a better home (e.g., `group_theory/group_action`)

jcommelin commented on src/group_theory/burnside.lean:
```diff
@@ -0,0 +1,153 @@
+Prove Burnside lemma (not due to Burnside): the number of orbits of a group action equals the average number of points fixed by an element of the group.
+-/
+
+import .group_action ..order.fixed_points
```
Please don't use relative paths for imports.

urkud replied:
Done.

jcommelin commented on src/group_theory/burnside.lean:
```diff
@@ -0,0 +1,153 @@
+import .group_action ..order.fixed_points
+
+/- TODO: What file is a good home for this? -/
+instance lattice.fixed_points_fintype
```
A good home would be somewhere after the definition of `lattice.fixed_points`.

urkud replied:
Then `order/fixed_points` will have to depend on `fintype`, and I'm not sure if it's desirable. BTW, is there any way to visualize `import` dependencies between files? How do you avoid circular dependencies?

ChrisHughes24 replied:
Putting this instance in `order/fixed_points` should mean that the `fintype` instance is synthesized automatically.
```lean
instance {α : Type*} [decidable_eq α] (f : α → α) : 
  decidable_pred (lattice.fixed_points f) :=
by dunfold lattice.fixed_points; apply_instance
```

urkud replied:
It seems that Lean eneters an infinite loop `subtype.fintype` ⟷ `set.decidable_mem_of_fintype` instead of finding this instance.

urkud replied:
@ChrisHughes24 In a new version of pull request, I both add an instance suggested by you, and an explicit instance in `group_theory/burnside` with a comment explaining that this is a temporary workaround.

jcommelin commented on src/group_theory/burnside.lean:
```diff
@@ -0,0 +1,153 @@
+
+variables {G : Type u} [group G] {α : Type v} [mul_action G α]
+
+@[simp]
```
If the attribute is short, we usually put it on the same line as `lemma ...`

urkud replied:
Done (everywhere).

jcommelin commented on src/group_theory/burnside.lean:
```diff
@@ -0,0 +1,153 @@
+variables {G : Type u} [group G] {α : Type v} [mul_action G α]
+
+@[simp]
+lemma smul_inv_left {g : G} {x : α} : g⁻¹ • g • x = x :=
```
These lemma's should go right after the definition of `mul_action`, in a section on `mul_action`s by groups. In that way they will be available to everyone who imports `group_action` in the future.

urkud replied:
Done.

jcommelin commented on src/group_theory/burnside.lean:
```diff
@@ -0,0 +1,153 @@
+  g' ∈ stabilizer G x ↔ (g * g' * g⁻¹ ∈ stabilizer G (g • x)) :=
+show _ • _ = _ ↔ _ • _ = _,
+begin
+  simp only [mul_smul,smul_inv_left],
```
```suggestion
  simp only [mul_smul, smul_inv_left],
```
Spaces after commas. Also elsewhere.

urkud replied:
Done (everywhere).

jcommelin commented on src/group_theory/burnside.lean:
```diff
@@ -0,0 +1,153 @@
+lemma smul_inv_right {g : G} {x : α} : g • g⁻¹ • x = x :=
+by rw [← mul_smul, mul_right_inv, one_smul]
+
+lemma stabilizer_orbit_iff {x : α} {g g' : G} :
```
I think I would call this `mem_stabilizer_iff` or `mem_stabilizer_iff_conj_mem`. I'm not sure... @ChrisHughes24 what would you call this?

urkud replied:
Renamed to `mem_stabilizer_smul_iff`.

jcommelin commented on src/group_theory/burnside.lean:
```diff
@@ -0,0 +1,153 @@
+  exact (bijective g).1.eq_iff
+end
+
+lemma stabilizer_orbit_inv_iff {x : α} {g g' : G} :
```
If it is decided to change the preceding name, a similar renaming should probably happen here.

urkud replied:
Renamed to `mem_stabilizer_self_iff_smul`.

jcommelin commented on src/group_theory/burnside.lean:
```diff
@@ -0,0 +1,153 @@
+end
+
+def equiv.stabilizer (g : G) (x : α) : stabilizer G x ≃ stabilizer G (g • x) :=
+begin
```
Can you use `refine_struct` to state the `to_fun` and `inv_fun` components in record notation?

jcommelin commented on src/group_theory/burnside.lean:
```diff
@@ -0,0 +1,153 @@
+end
+
+variables G α
+def orbits : Type v := quotient $ orbit_rel G α
```
I think everything up to this definition should just go into the `group_action.lean` file. It seems generally useful. (And thanks for writing those lemmas!)

urkud replied:
Done.

jcommelin commented on src/group_theory/burnside.lean:
```diff
@@ -0,0 +1,153 @@
+
+    Disjoint union of fixed points of all elements of the group is equivalent
+    to the product of the group and the orbits space. -/
+def burnside_equiv :
```
mathlib has (to mathematicians) somewhat strange naming conventions. In particular, we don't call lemma's that have a "specific name" (e.g. named after a person) by that name. We give them the descriptive name that we also give all the helper lemmas. It is explicitly encourage to give the "maths name" in the comments, like you already do.
So this should probably be called `sigma_fixed_points_equiv_group_prod_orbits`.

urkud replied:
I think, it would be nice to have a file in `docs/` with a list of named theorems, and their mathlib names. Or an attribute (e.g., @[title=Burnside Lemma]) with some "search by title" tactic.
I renamed my lemmas/defs, and will push it soon.

digama0 replied:
@urkud This is what the doc comments are for: cross referenced names, alternate names, and literature references (in addition to explanations of what the theorem is saying and relations to other formal theorems). Most theorems of course don't have all of these things, but that's the ideal.

The problem with calling a theorem "burnside's lemma" is that there may be many mathematically equivalent statements whose formal statements differ slightly, all of which would be called "burnside's lemma" by a mathematician shown the statement in isolation. The purpose of the formal name is to disambiguate these. If there is a reasonable argument that there is only one formal statement that could be given the name, and anyone thinking of the statement would call it "burnside" before a more direct name, then I would consider naming it something like `burnside`, but I don't think that applies in this case since there is another `burnside` right after this lemma, and it differs significantly in the formal statement.

urkud replied:
@digama0 Do I understand correctly that the best search engine for named theorems in mathlib is `git grep`? What do you think about adding named theorems to `docs/theories`? There we can have them with wikipedia/scolarpedia links and/or textbook references.

P.S.: Renamed `burnside_equiv` and `burnside` according to `mathlib` naming conventions.

digama0 replied:
@urkud Not quite. The `/-- -/` comments are machine readable, like Doxygen, although we currently aren't doing much with that information. But I don't see much of an advantage of another index in `docs` somewhere that can get out of date when there is already equivalent information interleaved with the definitions themselves, closer to the actual point of use.

It is possible to write a lean tactic that searches doc comments for a string, which should achieve your indexing goal.

urkud replied:
@digama0 As a potential user, I'd prefer to read some human-readable text about implementation of a specific theory in `mathlib` with links to specific facts than `/-- -/` comments. I mean, compare [sphinx](http://www.sphinx-doc.org/) based documentation with [doxygen](http://doxygen.nl/) or `coqdoc` output.
P.S.: I propose to close this branch of discussion as an off-topic in the context of the pull request. If I'll have some proposal, I'll come to Zulip, or open an RFC issue.

digama0 replied:
(I'm not blocking this PR based on this discussion, although I haven't reviewed the rest of it in depth. I defer to the other maintainers.) There are several sphinx documentation projects on lean + mathlib, but they usually develop into standalone books, e.g. Theorem Proving in Lean or Logic & Proof. My hope is to have `/-- -/` comments serve as source material for a doxygen/coqdoc style output. Some people use coqdoc for literate programming style (a.k.a 50/50 full narrative documentation), which is probably a bit too verbose for mathlib except maybe on isolated definitions, but I know that people have played with these levels of documentation in tutorials and such; @kbuzzard @PatrickMassot @NeilStrickland have used this for teaching purposes and may have more pointers.

urkud replied:
@digama0 Thank you for explanation. BTW, would it be easier to review this PR, if I split it into a few based on what parts of the source tree I modify?

jcommelin replied:
PRs of <250 lines are generally a lot easier to review. I'll leave it up to you whether you actually want to split it. I think this is now actually quite close to being ready for merging. @ChrisHughes24 What do you think?

jcommelin commented on src/group_theory/burnside.lean:
```diff
@@ -0,0 +1,153 @@
+    ...  ≃ G                                        : h₃
+end
+
+/-- Equivalence behind Burnside Lemma.
```
These docstrings show up in mouse-over tooltips in VScode. It is very good that you write them! However, I suggest that you compactify this a little bit. Maybe replace the `.` with a `:` and continue your explanation on the same line, instead of leaving an empty line.

urkud replied:
I'll do it. BTW, (a) are they integrated into Emacs `lean-mode`? (b) What do you think about adopting Doxygen style (brief description in the first line + detailed discussion after a blank like)? Tooltips can show the first line, and expand it on click/hover.

jcommelin replied:
@digama0 You're the boss concerning the style guide... what's the rule :wink: ?

digama0 replied:
I'm okay with longer comments; at least in vscode it's not a big deal to have a decent amount of text here, even up to 10-20 lines. The first line is not treated specially by emacs or vscode, but I think that summarizing in the first sentence is a good style regardless, with or without a paragraph break.

urkud replied:
I hope, comments look better now.

jcommelin commented on src/group_theory/burnside.lean:
```diff
@@ -0,0 +1,153 @@
+/-- Burnside's lemma: number of orbits of a group action equals the average number of fixed points of a group element.
+
+    For the case of infinite sets, see `burnside_equiv`. -/
+lemma burnside :
```
And this would be called `card_group_mul_card_orbits_eq`.

jcommelin commented on src/group_theory/group_action.lean:
```diff
@@ -131,20 +131,42 @@ variables {β}
+       ... ↔             h • b = g • b : by rw [← mul_action.mul_smul, mul_inv_cancel_left]
+       ... ↔             g • b = h • b : ⟨eq.symm, eq.symm⟩
+
+/-- Given an left coset w.r.t. the stabilizer of `b`, compute the image of `b`
```
```suggestion
/-- Given a left coset w.r.t. the stabilizer of `b`, compute the image of `b`
```

jcommelin commented on src/group_theory/group_action.lean:
```diff
@@ -131,20 +131,42 @@ variables {β}
+  (λ g h H, subtype.eq $ (stabilizer_coset_iff_image α b g h).1 H)
+
+/-- `orbit_of_quotient_stabilizer α b` sends `⟦x⟧` to `x • b`-/
+def orbit_of_quotient_stabilizer_spec (b : β) (x : α) :
```
This should be a `lemma`, and probably `@[simp]`.

urkud replied:
Done.

jcommelin commented on src/group_theory/group_action.lean:
```diff
@@ -131,20 +131,42 @@ variables {β}
+
+/-- Given an left coset w.r.t. the stabilizer of `b`, compute the image of `b`
+    under elements of this coset. -/
+def orbit_of_quotient_stabilizer (b : β) (x : quotient (stabilizer α b)) : orbit α b :=
```
Should this be `quotient_stabilizer_smul`? You can then use it to give the induced `mul_action` of the quotient on the orbit.

jcommelin reviewed (commented):
Thanks for this PR! I left some comments. Mostly about stylistic issues.

kckennylau commented on src/group_theory/group_action.lean:
```diff
@@ -131,20 +131,42 @@ variables {β}
+    w.r.t. the stabilizer of this point. For the computable part of this equivalence,
+    see `orbit_of_quotient_stabilizer`. -/
 noncomputable def orbit_equiv_quotient_stabilizer (b : β) :
   orbit α b ≃ quotient (stabilizer α b) :=
```
This is not just a bijection; this is an isomorphism of G-sets.

urkud replied:
I'd postpone the proof of this fact to another pull request.

NeilStrickland commented on src/group_theory/burnside.lean:
```diff
@@ -0,0 +1,153 @@
+def orbit_choice_spec {o : orbits G α} (x : {x // ⟦x⟧ = o}) :
+  orbit_choice x • o.out = x.val :=
+some_spec (quotient.exact $ x.2.trans o.out_eq.symm)
+
```
I would suggest that you give versions of the results below that depend on a variable that specifies an orbit choice.  Then you can have versions that use the choice given by `some (quotient.exact ...)`.  That will allow for computation in some concrete cases.

urkud replied:
1. Why would someone want to compute this? The particular bijection has no meaning at all.
2. The proof relies on a noncomputable group_equiv_quotient_times_subgroup, and I don't want to rewrite its proof.

urkud committed:
feat(data/quot): quot.map: act on non-id maps

urkud committed:
feat(algebra/group): define `units.conj`, and prove two trivial lemmas

Commit:
chore(group_theory/group_action): fix a typo in a comment

Co-Authored-By: Johan Commelin <johan@commelin.net>

urkud committed:
chore(algebra/group): def to_units

Sometimes `coe` fails to find the instance
```lean
group.has_lift : has_lift [group α] α → units α
```
so I define a named function.

urkud committed:
feat(data/equiv/basic): improve arrow_congr, define conj

- redefine `equiv.arrow_congr` without an enclosing `match`
- prove some trivial lemmas about `equiv.arrow_congr`
- define `equiv.conj`, and prove trivial lemmas about it

urkud committed:
feat(data/function/conjugate): (semi)conjugate functions

urkud committed:
feat(data/function/fixed_points): new file

- move definition from `order/fixed_points`
- prove some trivial lemmas, including correspondence between fixed
  points of (semi)conjugate maps
- adjust `group_theory/burnside`

urkud committed:
Merge branch 'burnside-lemma' of ssh://github.com/urkud/mathlib into burnside-lemma

urkud committed:
chore(data/quot): prove `⟦x⟧ = q ↔ x ≈ q.out`

This fact trivially follows from `quotient.eq` and `quotient.out_eq`,
but `▸` fails to do the trick, because it rewrites all occurences of `q`.

urkud committed:
chore(group_theory/group_action): rename a function

- `orbit_of_quotient_stabilizer` → quotient_stabilizer_smul as
  suggested by @jcommelin in
  https://github.com/leanprover-community/mathlib/pull/1082#discussion_r287551490

urkud committed:
chore(group_theory/group_action): use `α` and `β` from `variables` in defs

- `has_scalar` used `α` and ‵γ`

urkud committed:
chore(group_theory/burnside): use absolute imports

- see https://github.com/leanprover-community/mathlib/pull/1082#discussion_r287550819

urkud committed:
chore(group_theory/{group_action,burnside,sylow}): cleanup

- move some lemmas from `burnside` to `group_action`
- use better names
- cleanup proofs
- try to follow style guide
- use `mul_action.orbits` in `sylow`
- add explicit names to some instances
- move properties (symm, trans) of `orbit_rel` to lemmas

urkud committed:
chore(group_theory/group_action): fixup 9c8d37b4de64294ab29c17309c5d40c904b82132

- rename related lemmas/definitions as well
- fix def -> lemma

urkud committed:
chore(group_theory/group_action): add `@[simp]` marker

- suggested by @jcommelin https://github.com/leanprover-community/mathlib/pull/1082#discussion_r287551443

Commit:
Merge branch 'master' into burnside-lemma

Commit:
Merge branch 'master' into burnside-lemma

Commit:
Merge branch 'master' into burnside-lemma

ChrisHughes24 commented on src/data/function/conjugate.lean:
```diff
@@ -0,0 +1,120 @@
+
+namespace function
+  /-- Some sources require `g` to be surjective; if `g` is bijective, then `f₁` is conjugate to `f₂`. -/
+  def semiconj_by (g : α → β) : rel (α → α) (β → β) := λ f₁ f₂, ∀ x : α, g (f₁ x) = f₂ (g x)
```
We don't usually indent inside sections and namespaces.

ChrisHughes24 commented on src/data/function/fixed_points.lean:
```diff
@@ -0,0 +1,129 @@
+variables {α : Type u} {β : Type v}
+
+namespace function
+  def fixed_points (f : α → α) : set α := { x | f x = x }
```
This is already in `order.fixed_points`

urkud replied:
Being a fixed point of a map has nothing to do with order, so I moved it here, and `import`'ed this file in `order.fixed_points`.

ChrisHughes24 commented on src/data/function/conjugate.lean:
```diff
@@ -0,0 +1,120 @@
+    rfl
+  end semiconj_by
+
+  structure is_semiconj :=
```
Usually `is` indicates a `Prop`. Unless you are actually using this as a this as a type then it should be defined as an existence statement. I notice there is one lemma that refers to the `hom`, but this could be restated using the `semiconj_by_hom` relation.

ChrisHughes24 commented on src/data/function/conjugate.lean:
```diff
@@ -0,0 +1,120 @@
+  by rw [h, e.right_inv]
+
+  structure is_conj :=
+  (hom : α ≃ β)
```
There is already an `is_conj` in the root namespace I think, in `algebra.group`. Also same comment as with `is_semiconj`

urkud replied:
Is it bad to have both `_root_.is_conj` and `function.is_conj`?

urkud replied:
(Probably I'll rename it anyway due to the previous comment).

ChrisHughes24 commented on src/data/quot.lean:
```diff
@@ -77,6 +78,9 @@ noncomputable def quotient.out [s : setoid α] : quotient s → α := quot.out
 
 @[simp] theorem quotient.out_eq [s : setoid α] (q : quotient s) : ⟦q.out⟧ = q := q.out_eq
 
+theorem quotient.out_eq_iff [s : setoid α] (q : quotient s) (x : α) : ⟦x⟧ = q ↔ x ≈ q.out :=
```
This is `quotient.mk_eq_iff`

urkud replied:
`mk_eq_iff` doesn't mention `out`. Maybe, `mk_eq_iff_eqv_out`?

ChrisHughes24 commented on src/data/function/fixed_points.lean:
```diff
@@ -0,0 +1,129 @@
+  (is_conj.mk' e f).fixed_points_equiv
+
+  namespace perm
+    def fixed_points (e : equiv.perm α) : set α := function.fixed_points e
```
I would rather not have this defined separately for `perm`.

urkud replied:
I thought it would be convenient to write `e.fixed_points`. Of course, I can remove it.

ChrisHughes24 commented on src/group_theory/burnside.lean:
```diff
@@ -0,0 +1,89 @@
+    group and the orbits space. -/
+noncomputable def sigma_fixed_points_equiv_group_prod_orbits :
+  (Σ g : G, (to_perm G α g).fixed_points) ≃ (G × orbits G α) :=
+calc _ ≃ (Σ (o : orbits G α) (x : {x // ⟦x⟧ = o}), stabilizer G x.val) : sigma_fixed_points_equiv_sigma_stabilizer G α
```
lines too long.

ChrisHughes24 commented on src/group_theory/burnside.lean:
```diff
@@ -0,0 +1,89 @@
+
+/-- Burnside's lemma: number of orbits of a group action equals the average number of fixed points of a group element.
+    For the case of infinite sets, see `burnside_equiv`. -/
+lemma card_group_mul_card_orbits_eq : card G * card (orbits G α) = sum univ (λ g : G, card $ (to_perm G α g).fixed_points) :=
```
100 character lines. Also, there is a bit of ambiguity here given that there is already a `fixed_points` for a group action that means something different (points that are fixed by every element in the group). I think we should give these two things different names, probably the `mul_action` version should be renamed.

ChrisHughes24 replied:
The `fixed_points` in this definition is not the same as `mul_action.fixed_points`, and this could cause confusion. I think the name of `mul_action.fixed_points` should be changed.

ChrisHughes24 commented on src/group_theory/group_action.lean:
```diff
@@ -123,28 +150,68 @@ instance (b : β) : is_subgroup (stabilizer α b) :=
+
+/-- Given a left coset w.r.t. the stabilizer of `b`, compute the image of `b`
+    under elements of this coset. -/
+def quotient_stabilizer_smul (b : β) (x : quotient (stabilizer α b)) : orbit α b :=
```
This looks like it should be a `mul_action` instance

ChrisHughes24 replied:
This should be a `mul_action` instance.

ChrisHughes24 commented on src/group_theory/group_action.lean:
```diff
@@ -169,6 +236,59 @@ mul_action.comp_hom (quotient H) (subtype.val : I → α)
+
+variables (β)
+
+lemma conj_smul_funeq (a₁ : units α) (a₂ : α) :
```
Is this strictly necessary? Usually you can perform `rw`s using this with `conv` or `simp`

ChrisHughes24 commented on src/group_theory/group_action.lean:
```diff
@@ -169,6 +236,59 @@ mul_action.comp_hom (quotient H) (subtype.val : I → α)
+
+variables {β}
+
+lemma mem_stabilizer_smul_iff (u : units α) (x : α) (b : β) :
```
Usually the more complicated proposition is on the left of the `iff` in the statement. Also, the name `mem_stabilizer_smul_iff` suggests there is a `smul` in the left hand side of the statement, which there is not. I would suggest the name `conj_mem_stabilizer_smul_iff`

ChrisHughes24 replied:
I suggest changing the name of this to `conj_mem_stablizer_smul_iff`, and reversing the order of the `iff`. Usually the more complicated expression goes on the left hand side.

ChrisHughes24 commented on src/group_theory/group_action.lean:
```diff
@@ -169,6 +236,59 @@ mul_action.comp_hom (quotient H) (subtype.val : I → α)
+apply equiv.mem_fixed_points_of_conj
+end
+
+lemma mem_stabilizer_self_iff_smul (u : units α) (x : α) (b : β) :
```
This is `mem_stabilizer_smul_iff`

ChrisHughes24 commented on src/group_theory/group_action.lean:
```diff
@@ -169,6 +236,59 @@ mul_action.comp_hom (quotient H) (subtype.val : I → α)
+(mem_stabilizer_smul_iff u⁻¹ x (u.val • b)).trans
+  (eq.to_iff $ by congr; exact @smul_left_inv _ _ _ _ u b)
+
+def stabilizer_equiv_smul_units (u : units α) (b : β) :
```
This should be a `mul_equiv`.

ChrisHughes24 commented on src/group_theory/group_action.lean:
```diff
@@ -169,6 +236,59 @@ mul_action.comp_hom (quotient H) (subtype.val : I → α)
+
+variables [group α] [mul_action α β]
+
+def stabilizer_equiv_smul (a : α) (b : β) : stabilizer α b ≃ stabilizer α (a • b) :=
```
Again, this is a `mul_equiv
